File: LeetCode Solutions/Binary Tree/Symmetric Tree/Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if (not root):
            return True
        def isSymmetric(p,q) -> bool:
            if (not p and not q):
                return True
            if ((p and not q) or (not p and q) or (p.val != q.val)):
                return False
            if (not isSymmetric(p.left, q.right)):
                return False
            if (not isSymmetric(p.right, q.left)):
                return False
            return True
        return isSymmetric(root.left, root.right)

File: LeetCode Solutions/Binary Tree/Symmetric Tree/Symmetric Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == NULL){
            return true;
        }
        return isSymmetric(root,root);
    }
    bool isSymmetric(TreeNode* a, TreeNode* b){
        if (a == NULL && b == NULL){
            return true;
        }
        if (a != NULL && b != NULL && a->val == b->val){
            return isSymmetric(a->left, b->right) && isSymmetric(a->right, b->left);
        }
        return false;
    }
};